lib/syskit/log/datastore/dataset.rb

```diff
@@ -775,6 +776,98 @@ def each_task(
                     &block
                 )
             end
+
+            # Try converting the given value to the target using pocolog upgrade mechanisms
+            #
+            # "Upgrade" in this context is the mechanism through which data can be updated
+            # when types have changed over time, and/or because of change of architecture
+            # (for types that would have architecture-dependent features).
+            #
+            # Without any specific setup, the system will be able to handle change of layout
+            # (i.e. removal of fields, change of types providing the field new/old types
+            # are compatible, change of order of fields).
+            #
+            # For more complex upgrades, one has to provide custom conversion handlers. These
+            # handlers are Ruby files that register conversions on a
+            # {Pocolog::Upgrade::ConversionRegistry} object. See the rock-and-syskit
+            # documentation for more details
+            #
+            # @param [Typelib::Type] value the value to upgrade
+            # @param [Typelib::Type,Class<Typelib::Type>] target the conversion target,
+            #    either as a value or as a Typelib type.
+            # @return [Typelib::Type]
+            # @raise [Pocolog::Upgrade::InvalidCast] if the conversion is not possible
+            def upgrade_value_to(value, target, reference_time: interval_lg.first)
+                ops = upgrade_ops_for_target(value, target, reference_time)
+                return value if ops.identity?
+
+                target_value = upgrade_resolve_target_value(target)
+                ops.call(target_value, value)
+                target_value
+            end
+
+            # @private
+            #
+            # Calculate and caches the operations needed to upgrade the given value to
+            # `target`.
+            #
+            # @param [Typelib::Type] value the value to upgrade
+            # @param [Typelib::Type,Class<Typelib::Type>] target the conversion target,
+            #    either as a value or as a Typelib type.
+            # @return [Pocolog::Upgrade::Ops::Identity]
+            # @raise [Pocolog::Upgrade::InvalidCast] if the conversion is not possible
+            def upgrade_ops_for_target(value, target, reference_time)
+                unless datastore
+                    raise ArgumentError,
+                          "upgrade feature not available for datasets without "\
+                          "an associated datastore"
+                end
+
+                target_type = upgrade_resolve_target_type(target)
+                cache_key = [value.class, target_type, reference_time]
+
+                @upgrade_ops_for_target[cache_key] ||=
+                    Pocolog::Upgrade.compute(
+                        reference_time, value.class, target_type,
```

Comment:
Why Pocolog's upgrade graph needs a reference time? How this impact how a conversion from type A to B is done?

Reply:
This is for conversions from type A to type A, but fixing some issues in the data itself. Generally speaking, all converters have a "up to" date that limits their scope. Since converters are written in Ruby, it can be Time.now if said converter has no obvious due date.